# ITI5196 Assessment 2 Task2
#### Student Name: Eddy
#### Student ID: 33495068

Date: 13/11/2023

Version: 1.0

Environment: Python 3.10.11 and Jupyter notebook extension in VS code

Libraries used: please include the main libraries you used in your assignment here, e.g.,:
* pandas 
* re 
* numpy
* math
* mathplotlib.pyplot
* datetime
* collections
* difflib



<span style="color: red"> Note that this is a sample notebook only. You will need to fill in the proper markdown and code blocks. You might also want to make necessary changes to the structure to meet your own needs. It is important to make sure the logic of how you finish the assessment is clearly shown in this notebook! </span>

# Import Libraries

In [1]:
import pandas as pd
import math
import re
import numpy as np
import matplotlib.pyplot as plt

## Task 2. Integration

In this task, you are given a 2nd job advertisement dataset dataset2.csv. You will then integrate this dataset with the output from Task 1 (dataset1_solution.csv) and resolve any schema level and data level conflicts. 

<span style="color: red"> You might have complex notebook structure in this section, please feel free to create your own notebook structure. </span>

First we read both the dataset that's been wrangled and the second dataset

In [2]:
#Loading both dataset
dataset1 = pd.read_csv("dataset1_solution.csv")
dataset2 = pd.read_csv("dataset2.csv")

Next we EDA the second dataset first to get a feel of the dataset

In [3]:
#EDAing the second dataset
dataset2.info()
dataset2.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 334 entries, 0 to 333
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Job Title         334 non-null    object 
 1   Organisation      334 non-null    object 
 2   Location          334 non-null    object 
 3   Category          334 non-null    object 
 4   Salary per month  334 non-null    int64  
 5   Fraction          3 non-null      object 
 6   Contract Type     327 non-null    object 
 7   Views             334 non-null    float64
 8   OpeningDate       334 non-null    object 
 9   OpeningTime       334 non-null    object 
 10  ClosingDate       334 non-null    object 
 11  ClosingTime       334 non-null    object 
dtypes: float64(1), int64(1), object(10)
memory usage: 31.4+ KB


(334, 12)

In [4]:
#Showing the first 10 of second dataset
dataset2.head(10)


,Job Title,Organisation,Location,Category,Salary per month,Fraction,Contract Type,Views,OpeningDate,OpeningTime,ClosingDate,ClosingTime
0,Aviation loans administration,cer Financial,London,Finance and Accounting,2800,NaN,Contract,26899.161398,2013-10-06,00:00:00,2013-12-05,00:00:00
1,"Payroll Analyst City upto **** , ****",LMA Recruitment Ltd,London,Finance and Accounting,2917,NaN,Permanent,27998.181788,2012-10-03,12:00:00,2012-11-02,12:00:00
2,Investment Team Assistant for leading Private ...,Austin Andrew Ltd,London,Finance and Accounting,3750,NaN,Permanent,35955.060147,2012-01-01,00:00:00,2012-01-31,00:00:00
3,SWAPS COLLATERAL CONTROL OFFICER,Brian Durham Recruitment Services Limited,City,Finance and Accounting,3333,NaN,Permanent,31983.941441,2012-10-14,00:00:00,2012-11-13,00:00:00
4,Loans Administration Temp,cer Financial,London,Finance and Accounting,3280,NaN,Contract,31487.694078,2012-11-17,12:00:00,2013-01-16,12:00:00
5,Oversight and Compliance Manager Global Custo...,Citifocus Limited,The City,Finance and Accounting,3750,NaN,Permanent,36032.076862,2012-11-01,12:00:00,2012-12-31,12:00:00
6,ALM Actuary,Reed Insurance,South East England,Finance and Accounting,6250,NaN,Permanent,59953.789569,2012-03-22,00:00:00,2012-04-21,00:00:00
7,SUPPORT ANALYST Front Office,Newside Consulting Limited,London,Finance and Accounting,4667,NaN,Permanent,44848.708848,2013-03-09,12:00:00,2013-05-08,12:00:00
8,Forex Sales Manager,LMA Recruitment Ltd,London,Finance and Accounting,4583,NaN,Permanent,44035.320755,2013-11-26,12:00:00,2013-12-26,12:00:00
9,Management Accountant Commodity Trading,LMA Recruitment Ltd,London,Finance and Accounting,4792,NaN,Permanent,46031.453138,2013-06-21,15:00:00,2013-09-19,15:00:00


From looking at the dataset it's clear that this dataset doesn't match the schema that were given in the global schema. Therefore, we will look at each column problem and fix it from there. First off the missing Id column.

# Id

As mentioned before, the column Id is missing. From the global schema, the Id column is 8 digit id of the job advertisement. From what was observed in the first dataset. They are all unique therefore they can serve as the Primary Key for both dataset. Because of that, when making the Id column for this dataset it's also must be a unique 8 digit dataset. To make sure it's unique, we check the maximum number from the first dataset then increase it incrementaly for the 2nd dataset

In [5]:
#Code to show the maximum id for dataset 1
max(dataset1['Id'])

72705504

As we can see the max Id is 72705504 so we increase it from there for dataset2

In [6]:
#Code to insert the id column increasing from 72705505. And to make sure the name match the global schema Id
dataset2.insert(0, 'Id', range(72705505, 72705505 + len(dataset2)))

Next is the Job Title Column

# Job Title

First we skim the column with for loop

In [7]:
#Skimming the column
for x,y in dataset2['Job Title'].value_counts().to_dict().items():
    print(x,y)

Aviation loans administration 1
Lease Accountant 1
Assistant Assurance Manager 1
Cost Accountant  Supply Chain Finance 1
Rent Accountant / Property Accountant 1
Product Support Analyst 1
OTC Client Services 1
PPI Quality Assurer 1
Senior Financial Accountant Insurance 1
IRD and Commodities Confirmations Specialists 1
Credit Risk Reporting Manager / Assistant Manager 1
Snr Relationship Executive  Fund Management  Fluent German 1
Business Analyst  BAU and Change 1
Business Analyst  OTC Pricing 1
Compliance Associate 1
Credit Risk Senior Analyst 1
Property Finance Manager 1
Leading Gloabl Consultancy  Capital Markets Consultants 1
Impairment and Capital Manager 1
Internal Auditor and Senior Auditor 1
KYC Remediation 1
Sales Administrator/ Client Service Associate 1
Financial Crime Analyst (Compliance) 1
Associate Director Product and Support  (Cash Management) 1
IFRS Accountant 1
Operational Due Diligence (ODD) Analyst 1
Physical Oil Settlements (Pricing ) Manager 1
Legal Compliance Regul

The most obvious mistake is that the column names didn't match the global schema. The data are also not yet normalized therefore we normalize the data according to what has been done in task 1 and change the column name

In [8]:
#Change the column name from job title to Title
dataset2 = dataset2.rename(columns={'Job Title': 'Title'})

#Normalize the data like in dataset1

#Creating a function to normalize the title column
def normalizetitle(title):
    # normalize to upper case letters
    title = title.upper()
    # change AND to &
    title = title.replace(" AND ", " & ")
    # replace multiple spaces with a single space, also trim spaces on both side
    title = re.sub( '\s+', ' ', title).strip()
    return title

#Applying the function to the column
dataset2['Title'] = dataset2.Title.apply(lambda x: normalizetitle(x))

We're done with Title column, next Organisation column

# Organisation

First we skim the column with for loop

In [9]:
#Skimming the column
for x,y in dataset2['Organisation'].value_counts().to_dict().items():
    print(x,y)

LMA Recruitment Ltd 19
Oliver James Associates Limited 15
Brian Durham Recruitment Services Limited 12
Ambition Europe Limited 12
cer Financial 11
Morgan McKinley Group Limited 11
Kennedy Pearce Consulting Ltd 10
Goodman Masson Recruitment Services Ltd 10
Randstad Financial Professional Ltd 10
Badenoch Clark Ltd 9
Hays Financial Markets 9
Prime Personnel Services Ltd 9
Idex Consulting LLP 8
The Oakland Partnership Limited 8
Black Swan Associates Limited 8
Clark James Ltd 7
Twenty Recruitment Ltd 7
C.K.R. Recruitment Limited 7
Incite Solutions Ltd 6
Austin Andrew Ltd 6
Reed Finance 6
Austin Benn 6
Argyll Scott International Limited 5
Alexander Lloyd 4
Venn Group Limited 4
Mayford James Limited 4
Montpellier Resourcing Ltd 3
Livingston Edwards Ltd 3
James Associates Recruitment Ltd 3
Employment Specialists Ltd 3
Insight Recruitment Solutions Ltd 3
Bruin Financial Limited 3
Walker Hamill Ltd 3
Marks Sattin  Ltd 3
Maldon Partners Ltd 3
Harrison Holgate 3
Monument Financial Recruitment Ltd 

Again the most obvious problem is the column names didn't match the global schema. There's also some inconsistency and upper and lower cases problem. Therefore we change the column name and normalize the data like in task 1

In [10]:
#Change the column name from job title to Title
dataset2 = dataset2.rename(columns={'Organisation': 'Company'})

#Normalize the data like in dataset1

# the following function is the clean and normalize the company field.
def normalizecompany(company):
    if pd.isnull(company):
        return company
    else:
        # normalize to upper case letters
        company = company.upper()
        # replace multiple spaces with a single space, also trim spaces on both side
        company = re.sub( '\s+', ' ', company).strip()
        # remove special character at the end
        company = re.sub(r"\W+$", "", company)
        # replace LIMITED with LTD
        company = company.replace("LIMITED", "LTD")
        # replace " AND " with " & " for processing 
        company = company.replace(" & ", " AND ")
        # remove all special characters except space and dot
        company = re.sub(r'([^\w\s\.]|_)', '', company)
        # change " AND " back to " & " 
        company = company.replace(" AND ", " & ")
        return company

#Applying the function to the column
dataset2['Company'] = dataset2.Company.apply(lambda x: normalizecompany(x))

Organisation column is done, next is location

# Location

First we skim the column with for loop

In [11]:
#Skimming the column
for x,y in dataset2['Location'].value_counts().to_dict().items():
    print(x,y)

London 158
The City 92
UK 25
City 13
South East England 5
West End 5
South West London 3
Central London 3
Crawley 2
Hertfordshire 2
Birmingham 2
Croydon 2
East Midlands 1
Reading 1
North London 1
Guildford 1
West Sussex 1
South East London 1
Basingstoke 1
Buckinghamshire 1
Southampton 1
Peterborough 1
Bristol 1
North West England 1
Farnham 1
West London 1
Chelmsford 1
Fareham 1
Hampshire 1
Berkshire 1
Redhill 1
Dorking 1
Cheltenham 1
Bath 1


Doesn't look like there's obvious mistake with the column. So we just normalize the column and move on

In [12]:
#Defining function to normalize the location by uppercasing the location and deleting spaces + trailing spaces

def normalizelocation(location):
        #We know that location column doesn't have missing values do no need to deal with values
        # normalize to upper case letters
        location = location.upper()
        # replace multiple spaces with a single space, also trim spaces on both side
        return location
dataset2['Location'] = dataset2.Location.apply(lambda x: normalizelocation(x))

Next is category

# Category

First we skim the column with for loop

In [13]:
#Skimming the column
for x,y in dataset2['Category'].value_counts().to_dict().items():
    print(x,y)

Finance and Accounting 319
Information Technology 12
PR, Advertising and Marketing 3


First, the column name already match the global schema. However, the data didn't match what has been cleaned in dataset 1 therefore it is normalized and matched to the dataset 1

In [14]:
#Defining function to normalize the category by uppercasing the category and deleting spaces + trailing spaces

def normalizecategory(category):
    #function to skip null is not needed because we known there's no null in this column    
    # normalize to upper case letters        
    category = category.upper()
    # replace " AND " with " & "
    category = category.replace(" AND ", " & ")
    return category

#Applying the function
dataset2['Category'] = dataset2.Category.apply(lambda x: normalizecategory(x))

#Code to correct the incorrect category names by replacing it with the correct names

dataset2.Category.replace({'FINANCE & ACCOUNTING': 'ACCOUNTING & FINANCE JOBS','INFORMATION TECHNOLOGY': 'IT JOBS',
                          'PR, ADVERTISING & MARKETING': 'PR, ADVERTISING & MARKETING JOBS'}, inplace=True)

Next is the Salary per month Column

# Salary Per Month

First we skim the column with for loop

In [15]:
#Skimming the column
for x,y in dataset2['Salary per month'].value_counts().to_dict().items():
    print(x,y)

3333 19
3750 18
4583 15
4167 14
3958 13
3125 12
3542 12
5000 11
4375 11
2917 11
5417 10
6250 10
5833 9
2500 9
2292 8
2083 7
4792 7
5208 6
2708 6
7500 6
7083 5
1875 5
6667 4
3167 4
2375 4
4000 4
2000 4
5625 3
2333 3
2800 3
5500 3
2667 3
2417 3
7917 3
1667 3
2250 2
3000 2
3833 2
6042 2
1750 2
2208 2
4708 2
2625 2
3208 2
4667 2
6400 2
3458 2
1708 1
1792 1
5750 1
4200 1
3417 1
1458 1
1625 1
3200 1
2583 1
1250 1
3060 1
3280 1
2167 1
4480 1
3583 1
5958 1
6458 1
5542 1
3250 1
1520 1
2833 1
3021 1
4292 1
1167 1
3760 1
2146 1
2125 1
3500 1
1417 1
2400 1
3800 1
4500 1
6700 1
6500 1
2880 1
1500 1
3520 1
6875 1
1583 1
3600 1
3440 1
1560 1


It's clear that there are problems with this column. First the name didn't match the global schema. Second the salary is in month format when it should be in Annual format which menas it should be / year so both problems are fixed. Also from the EDA at the start, the data format is integer when it should be float so it's also fixed.

In [16]:
#Change the column name from Salary per month to Salary
dataset2 = dataset2.rename(columns={'Salary per month': 'Salary'})

#Changing the Salary from per month to per year and coverting it to float
dataset2['Salary'] = dataset2.Salary.apply(lambda x: float(x * 12))

Next is Fraction column

# Fraction

First we skim the column with for loop

In [17]:
#Skimming the column
for x,y in dataset2['Fraction'].value_counts().to_dict().items():
    print(x,y)

Part Time 3


From the global schema, there's no column named fraction but looking at the code output. It seems that this column should be ContractType column. Also from the EDA earlier and the output of the code. It's clear that there are missing data in the column. As with task 1, the NA values is changed to non_specified

In [18]:
#Renaming the column name to ContractType
dataset2 = dataset2.rename(columns={'Fraction': 'ContractType'})

#Renaming the data to the correct values
dataset2.ContractType.replace('Part Time', 'part_time', inplace=True)
dataset2.ContractType.fillna('non-specified', inplace=True)

Next is Contract Type

# Contract Type

First we skim the column with for loop

In [19]:
#Skimming the column
for x,y in dataset2['Contract Type'].value_counts().to_dict().items():
    print(x,y)

Permanent 272
Contract 55


From the output, it seems this column should be ContractTime column in the global schema. From the EDA earlier we also know that there are missing data in this column, so those missing datas are replaced with non_specified according to the global schema

In [20]:
#Renaming the column name to ContractType
dataset2 = dataset2.rename(columns={'Contract Type': 'ContractTime'})

#Renaming the data to the correct values
dataset2.ContractTime.replace('Permanent', 'permanent', inplace=True)
dataset2.ContractTime.replace('Contract', 'contract', inplace=True)
dataset2.ContractTime.fillna('non-specified', inplace=True)

Next is the Views column

# Views

First we skim the column with for loop

In [21]:
#Skimming the column
for x,y in dataset2['Views'].value_counts().to_dict().items():
    print(x,y)

26899.161397970605 1
34007.41425216036 1
32004.213169285074 1
32810.31319769833 1
36039.74406175828 1
36027.62141528344 1
38001.10418790716 1
40344.636603278544 1
39983.05847559989 1
37967.76076561211 1
40025.07772236232 1
42041.20743707225 1
47972.54467749888 1
51984.167476841685 1
53956.96161184573 1
47989.05455043044 1
51959.76603535924 1
45212.29328021178 1
55204.07194244721 1
32037.13195359884 1
38437.20450959419 1
16810.384167167067 1
27990.19624434398 1
21983.72566644667 1
41959.82348739989 1
37962.57503727589 1
47977.78433261791 1
52797.17945846199 1
76049.59185261675 1
15563.021488916258 1
21561.473589353387 1
22361.472779541404 1
23956.584412251315 1
23972.547816552888 1
22827.084355378287 1
25979.712139727617 1
14025.204735254758 1
28012.42844325566 1
25651.81493778961 1
31984.803110371373 1
55975.82372712514 1
17187.64362867477 1
36005.75781681212 1
34029.464470547144 1
58040.01847403868 1
12000.821499404925 1
19158.725338249973 1
19979.926577643 1
27222.04941244969 1
30441

From the global schema, we knew that the name is wrong since it should be NumberOfViews. From the output of the code, the data seems to be in float type whereas from the global schema it should be in integer. Therefore the numbers are rounded down and converted to integer

In [22]:
#Renaming the column to NumberOfViews
dataset2 = dataset2.rename(columns={'Views': 'NumberOfViews'})

#Converting number of views to integer and rounding up the number to upper number.
dataset2.NumberOfViews = dataset2.NumberOfViews.round(0).astype('Int64')

Next is the OpeningDate and OpeningTime

# OpeningDate and OpeningTime

First we skim the column with for loop

In [23]:
#Skimming the OpeningDate column
for x,y in dataset2['OpeningDate'].value_counts().to_dict().items():
    print(x,y)

2012-03-04 4
2013-01-23 3
2013-08-29 3
2012-11-23 3
2012-01-25 3
2012-09-02 3
2013-04-26 3
2012-04-15 3
2012-08-22 3
2012-12-01 2
2012-05-09 2
2013-07-25 2
2013-05-11 2
2012-09-12 2
2012-10-06 2
2012-05-26 2
2013-01-03 2
2013-06-26 2
2012-12-13 2
2012-11-12 2
2013-08-10 2
2013-11-29 2
2013-04-21 2
2013-06-10 2
2013-10-28 2
2012-02-13 2
2013-04-07 2
2012-06-11 2
2012-05-30 2
2013-10-02 2
2013-02-24 2
2013-07-09 2
2013-04-01 2
2013-01-25 2
2013-05-22 2
2013-02-28 2
2012-11-06 2
2012-07-18 2
2013-10-24 2
2012-05-11 2
2013-10-25 2
2013-02-10 2
2013-01-05 2
2012-11-27 2
2012-11-17 2
2012-10-14 2
2013-02-02 2
2012-01-31 2
2013-01-06 2
2012-03-14 2
2013-02-11 2
2013-04-02 2
2013-11-17 2
2013-11-13 1
2012-07-06 1
2012-03-07 1
2013-04-14 1
2012-04-06 1
2013-04-08 1
2013-02-09 1
2012-01-12 1
2012-01-04 1
2013-10-06 1
2013-05-14 1
2012-01-30 1
2013-04-09 1
2013-09-18 1
2013-06-14 1
2012-08-08 1
2012-08-17 1
2012-11-16 1
2012-01-14 1
2013-03-31 1
2013-09-16 1
2012-07-29 1
2013-06-09 1
2012-03-29 1

In [24]:
#Skimming the OpeningTime column
for x,y in dataset2['OpeningTime'].value_counts().to_dict().items():
    print(x,y)

15:00:00 116
12:00:00 115
00:00:00 103


Few problems, from the global schema opening date and opening time are combined into one column of OpenDate. The format for it is dateThour and it's in string format. Therefore, both column are combined into one OpenDate column with T in the middle. Then the former 2 columns are dropped

In [25]:
#combining opening date and opening time to opendate
dataset2['OpenDate'] = dataset2['OpeningDate'].str.split('-', expand=False).str[0:3].apply(''. join) + 'T' + dataset2['OpeningTime'].str.split(':', expand=False).str[0:3].apply(''. join)

#dropping the OpeningTime and OpeningDate column
dataset2 = dataset2.drop(columns=['OpeningTime', 'OpeningDate'])

Next is the ClosingTime and ClosingDate

# ClosingTime and ClosingDate

First we skim the column with for loop

In [26]:
#Skimming the ClosingDate column
for x,y in dataset2['ClosingDate'].value_counts().to_dict().items():
    print(x,y)

2013-03-12 3
2012-09-16 3
2013-09-12 3
2013-08-09 3
2013-12-17 2
2012-09-17 2
2012-04-24 2
2013-09-23 2
2012-06-14 2
2013-01-12 2
2013-04-06 2
2014-01-01 2
2012-06-28 2
2013-07-20 2
2013-10-24 2
2012-05-13 2
2013-06-21 2
2012-07-25 2
2012-05-17 2
2013-11-01 2
2013-03-24 2
2012-05-03 2
2013-02-22 2
2012-05-06 2
2013-07-09 2
2013-03-28 2
2012-11-12 2
2012-11-04 2
2012-12-05 2
2012-12-12 2
2013-07-01 2
2013-04-03 2
2013-05-26 2
2013-03-13 2
2013-06-08 2
2012-06-29 2
2013-01-26 2
2012-04-13 2
2012-04-03 2
2012-12-06 2
2013-12-02 2
2013-11-27 2
2013-11-20 2
2013-03-30 2
2013-08-23 2
2014-01-23 2
2013-11-12 2
2012-04-30 2
2013-05-25 2
2012-10-12 2
2012-10-21 2
2013-12-26 2
2012-12-31 2
2013-01-16 2
2012-11-13 2
2013-09-28 2
2014-02-15 1
2013-04-10 1
2012-09-04 1
2013-06-07 1
2012-12-16 1
2012-03-04 1
2012-02-11 1
2013-04-28 1
2013-05-28 1
2013-02-08 1
2012-09-20 1
2012-09-07 1
2013-05-30 1
2012-08-27 1
2013-10-04 1
2012-04-28 1
2014-01-12 1
2012-05-23 1
2014-01-14 1
2012-07-14 1
2012-06-19 1

In [27]:
#Skimming the OpeningTime column
for x,y in dataset2['ClosingTime'].value_counts().to_dict().items():
    print(x,y)

15:00:00 116
12:00:00 115
00:00:00 103


Few problems, from the global schema closing date and Closing time are combined into one column of CloseDate. The format for it is dateThour and it's in string format. Therefore, both column are combined into one CloseDate column with T in the middle. Then the former 2 columns are dropped

In [28]:
#combining closing date and closing time to closedate
dataset2['CloseDate'] = dataset2['ClosingDate'].str.split('-', expand=False).str[0:3].apply(''. join) + 'T' + dataset2['ClosingTime'].str.split(':', expand=False).str[0:3].apply(''. join)

#dropping the OpeningTime and OpeningDate column
dataset2 = dataset2.drop(columns=['ClosingDate', 'ClosingTime'])

It's done, next we're checking if there's any problem

In [29]:
dataset2.info()
dataset2.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 334 entries, 0 to 333
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             334 non-null    int64  
 1   Title          334 non-null    object 
 2   Company        334 non-null    object 
 3   Location       334 non-null    object 
 4   Category       334 non-null    object 
 5   Salary         334 non-null    float64
 6   ContractType   334 non-null    object 
 7   ContractTime   334 non-null    object 
 8   NumberOfViews  334 non-null    Int64  
 9   OpenDate       334 non-null    object 
 10  CloseDate      334 non-null    object 
dtypes: Int64(1), float64(1), int64(1), object(8)
memory usage: 29.2+ KB


,Id,Title,Company,Location,Category,Salary,ContractType,ContractTime,NumberOfViews,OpenDate,CloseDate
0,72705505,AVIATION LOANS ADMINISTRATION,CER FINANCIAL,LONDON,ACCOUNTING & FINANCE JOBS,33600.0,non-specified,contract,26899,20131006T000000,20131205T000000
1,72705506,"PAYROLL ANALYST CITY UPTO **** , ****",LMA RECRUITMENT LTD,LONDON,ACCOUNTING & FINANCE JOBS,35004.0,non-specified,permanent,27998,20121003T120000,20121102T120000
2,72705507,INVESTMENT TEAM ASSISTANT FOR LEADING PRIVATE ...,AUSTIN ANDREW LTD,LONDON,ACCOUNTING & FINANCE JOBS,45000.0,non-specified,permanent,35955,20120101T000000,20120131T000000
3,72705508,SWAPS COLLATERAL CONTROL OFFICER,BRIAN DURHAM RECRUITMENT SERVICES LTD,CITY,ACCOUNTING & FINANCE JOBS,39996.0,non-specified,permanent,31984,20121014T000000,20121113T000000
4,72705509,LOANS ADMINISTRATION TEMP,CER FINANCIAL,LONDON,ACCOUNTING & FINANCE JOBS,39360.0,non-specified,contract,31488,20121117T120000,20130116T120000


# SourceName

The source name column is actually missing from the second dataset. The problem is that we can't really deduct where the source of each job offer came from. Therefore it's left as missing values for the SourceName column. When the dataset are both merged, it will automatically update for dataset2 and input it as missing values

Finally both dataset from task 1 and dataset2 are merged together

# Merging

Merging both dataset together

In [30]:
#Function to merge both dataset1 and dataset 2 into one dataset
merge = pd.concat([dataset1,dataset2], axis=0)

#Checking if the new dataset has both dataset
merge.info()
merge.tail(10)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 55499 entries, 0 to 333
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             55499 non-null  int64  
 1   Title          55499 non-null  object 
 2   Location       55499 non-null  object 
 3   Company        55499 non-null  object 
 4   ContractType   55499 non-null  object 
 5   ContractTime   55499 non-null  object 
 6   Category       55499 non-null  object 
 7   Salary         55499 non-null  float64
 8   OpenDate       55499 non-null  object 
 9   CloseDate      55499 non-null  object 
 10  SourceName     55165 non-null  object 
 11  NumberOfViews  55499 non-null  Int64  
dtypes: Int64(1), float64(1), int64(1), object(9)
memory usage: 5.6+ MB


,Id,Title,Location,Company,ContractType,ContractTime,Category,Salary,OpenDate,CloseDate,SourceName,NumberOfViews
324,72705829,CAPITAL MARKETS MANAGEMENT ACCOUNTANT,LONDON,ALBANY BECK CONSULTING LTD,non-specified,permanent,ACCOUNTING & FINANCE JOBS,54996.0,20121117T150000,20130116T150000,NaN,44005
325,72705830,COVERED BONDS CREDIT ANALYST,LONDON,RANDSTAD FINANCIAL PROFESSIONAL LTD,non-specified,permanent,ACCOUNTING & FINANCE JOBS,66504.0,20131229T150000,20140329T150000,NaN,53171
326,72705831,SENIOR BUSINESS BANKING RELATIONSHIP MANAGER E...,NORTH LONDON,THE OAKLAND PARTNERSHIP LTD,non-specified,permanent,ACCOUNTING & FINANCE JOBS,50004.0,20130210T120000,20130312T120000,NaN,39982
327,72705832,LOANS ANALYST,LONDON,BLAYNEY PERSONNEL LTD,non-specified,permanent,ACCOUNTING & FINANCE JOBS,42504.0,20131025T120000,20131224T120000,NaN,34047
328,72705833,NEWLY QUALIFIED ASSOCIATE,LONDON,REED FINANCE,non-specified,permanent,ACCOUNTING & FINANCE JOBS,42504.0,20120501T150000,20120630T150000,NaN,33986
329,72705834,FINANCIAL CONTROL OFFICER REGULATORY REPORTING,THE CITY,HAYS FINANCIAL MARKETS,non-specified,permanent,ACCOUNTING & FINANCE JOBS,39996.0,20130531T000000,20130829T000000,NaN,31991
330,72705835,MANAGEMENT ACCOUNTANT LONDON,THE CITY,MORGAN MCKINLEY GROUP LTD,non-specified,contract,ACCOUNTING & FINANCE JOBS,39000.0,20121111T120000,20121125T120000,NaN,31222
331,72705836,RISK MANAGER FSA / REGULATORY FRAMEWORKS,BATH,INCITE SOLUTIONS LTD,non-specified,permanent,ACCOUNTING & FINANCE JOBS,37500.0,20130326T150000,20130525T150000,NaN,30003
332,72705837,RELATIONSHIP / SALES EXECUTIVE MULTILINGUAL,LONDON,C.K.R. RECRUITMENT LTD,non-specified,permanent,"PR, ADVERTISING & MARKETING JOBS",30000.0,20130714T150000,20130912T150000,NaN,23959
333,72705838,CORPORATE TAX COMPLIANCE CLIENT SERVICE ADMINI...,UK,RANDSTAD FINANCIAL PROFESSIONAL LTD,non-specified,contract,ACCOUNTING & FINANCE JOBS,18720.0,20120218T150000,20120303T150000,NaN,14956


Looks like both dataset are merged successfuly, the first dataset has 55165 data after deleting duplicates and second dataset has 334 data. Adding both of them together we got 55499 except for the SourceName since the second dataset SourceName is left as NaN. We can also see the NaN values for the second dataset that we mentioned before that should be left as missing values. Next we need to check if the merging resulted in data conflicts

# DUPLICATES

There's a possibility that there are data conflicts in form of duplicates therefore we should check the combined dataframe again for duplicates. Like in task 1, we use 3 columns to search for duplicates which are Title, Location, and Category.

- Title is choosen because it's more likely that the name could be the same
- OpenDate and CloseDate is choosen because if the duplicated rows has the same name and has the same OpenDate and CloseDate with each other,
it's more likely that it's a duplication

In [31]:
#Checking the merged dataset for duplicates based by Title, OpenDate, and CloseDate
merge[merge.duplicated(['Title', 'OpenDate', 'CloseDate'], keep=False)].sort_values(['Title'])

,Id,Title,Location,Company,ContractType,ContractTime,Category,Salary,OpenDate,CloseDate,SourceName,NumberOfViews
22758,68840315,ACCOUNTS PAYABLE / COMPLIANCE / ADMINISTRATOR,LONDON,PRIME PERSONNEL SERVICES LTD,non_specified,permanent,ACCOUNTING & FINANCE JOBS,28500.0,20130106T000000,20130205T000000,jobs.catererandhotelkeeper.com,31851
155,72705660,ACCOUNTS PAYABLE / COMPLIANCE / ADMINISTRATOR,LONDON,PRIME PERSONNEL SERVICES LTD,non-specified,permanent,ACCOUNTING & FINANCE JOBS,28500.0,20130106T000000,20130205T000000,NaN,22758
6331,66894092,ADVANCED DATA ANALYTIC MANAGERS,THE CITY,OLIVER JAMES ASSOCIATES LTD,non_specified,permanent,IT JOBS,57500.0,20130106T120000,20130406T120000,staffnurse.com,34793
195,72705700,ADVANCED DATA ANALYTIC MANAGERS,THE CITY,OLIVER JAMES ASSOCIATES LTD,non-specified,permanent,IT JOBS,57504.0,20130106T120000,20130406T120000,NaN,45973
12470,68218723,ASSISTANT ACCOUNTANT FINANCIAL SERVICES EXPERI...,LONDON,MALDON PARTNERS LTD,non_specified,permanent,ACCOUNTING & FINANCE JOBS,27500.0,20131105T000000,20140104T000000,cwjobs.co.uk,31851
...,...,...,...,...,...,...,...,...,...,...,...,...
2420,62008199,SUPPORT ANALYST FRONT OFFICE,LONDON,NEWSIDE CONSULTING LTD,non_specified,permanent,ACCOUNTING & FINANCE JOBS,56000.0,20130309T120000,20130508T120000,careworx.co.uk,31851
53335,72460519,TRADE FINANCE OFFICER,LONDON,CER FINANCIAL,non_specified,contract,ACCOUNTING & FINANCE JOBS,46000.0,20120930T000000,20121030T000000,jobs.guardian.co.uk,31851
142,72705647,TRADE FINANCE OFFICER,LONDON,CER FINANCIAL,non-specified,contract,ACCOUNTING & FINANCE JOBS,45996.0,20120930T000000,20121030T000000,NaN,36792
28262,69266259,TRADE STRUCTURED FINANCE MANAGER COMMODITY TRADER,LONDON,LMA RECRUITMENT LTD,non_specified,permanent,ACCOUNTING & FINANCE JOBS,72500.0,20131117T120000,20140215T120000,cv-library.co.uk,31851


As we can see from the ouptut, there are a lot of duplicates in the merged dataset. I'm assuming that the data from the first dataset is the actual data and the data from the second dataset is the duplicates therefore, the second data is dropped.

In [32]:
#Code to drop the duplicates while keeping the first one
merge = merge.drop_duplicates(subset=['Title', 'OpenDate', 'CloseDate'], keep='first')

## The Global Key
Write down the identified global key:

As mentioned in the duplicates section, we use 3 columns to identify the duplicates therefore the Global Key are

**| Title  | OpenDate | CloseDate  |**

## Saving data
Save the integrated data

In [33]:
# code to save output data
merge.to_csv("dataset_integrated.csv",index = False)

## Summary
Give a short summary and anything you would like to talk about assessment 2 part 2 here.

For the task 2, the dataset is checked with the global schema and dataset 1 then the schema and data conflict is resolved. After that both dataset is merged, then checked again for duplicates. After duplicates are deleted then it's saved into dataset_integrated.csv